scylla-cdc-replicator/src/main.rs

```diff
@@ -1,29 +1,93 @@
+use futures_util::StreamExt;
+use tokio::select;
+use tokio::time::sleep;
+
+use crate::replicator::Replicator;
+
+#[derive(Parser)]
+struct Args {
+    /// Keyspace name
+    #[clap(short, long)]
+    keyspace: String,
+
+    /// Table names provided as a comma delimited string
+    #[clap(short, long)]
+    table: String,
+
+    /// Address of a node in source cluster
+    #[clap(short, long)]
+    source: String,
+
+    /// Address of a node in destination cluster
+    #[clap(short, long)]
+    destination: String,
+
+    /// Window size in seconds
+    #[clap(long, default_value_t = 60.)]
+    window_size: f64,
+
+    /// Safety interval in seconds
+    #[clap(long, default_value_t = 30.)]
+    safety_interval: f64,
+
+    /// Sleep interval in seconds
+    #[clap(long, default_value_t = 10.)]
+    sleep_interval: f64,
+
+    /// Start datetime as RFC 3339 formatted string
+    #[clap(long)]
+    start_datetime: Option<String>,
+}
+
 #[tokio::main]
-async fn main() -> Result<(), Box<dyn Error>> {
-    println!("Hello");
-    tokio::time::sleep(Duration::from_secs(1)).await;
-
-    // Create a factory to remove the warning about unused new() function.
-    let _factory = ReplicatorConsumerFactory::new(
-        Arc::new(
-            SessionBuilder::new()
-                .known_node("127.0.0.1:9042")
-                .build()
-                .await?,
-        ),
-        "ks".to_string(),
-        "t".to_string(),
-    );
-
-    println!("Replicator");
-    Ok(())
+async fn main() -> anyhow::Result<()> {
+    let args = Args::parse();
+    let tables: Vec<_> = args.table.split(',').map(|s| s.to_string()).collect();
+    let start_timestamp = chrono::Duration::milliseconds(match args.start_datetime {
+        Some(s) => chrono::DateTime::parse_from_rfc3339(&s)?.timestamp_millis(),
+        None => chrono::Local::now().timestamp_millis(),
+    });
+    let sleep_interval = Duration::from_secs_f64(args.sleep_interval);
+
+    let (mut replicator, mut handles) = Replicator::new(
+        args.source,
+        args.keyspace.clone(),
+        tables.clone(),
+        args.destination,
+        args.keyspace,
+        tables,
+        start_timestamp,
+        Duration::from_secs_f64(args.window_size),
+        Duration::from_secs_f64(args.safety_interval),
+        sleep_interval,
+    )
+    .await?;
+
+    loop {
+        select! {
+            _ = tokio::signal::ctrl_c() => {
+                println!("Shutting down...");
+                replicator.stop();
+                while handles.next().await.is_some() {}
+                return Ok(())
+            }
+            res = handles.next(), if !handles.is_terminated() => {
+                let res = res.unwrap();
+                if res.is_err() {
+                    replicator.stop();
+                    while handles.next().await.is_some() {}
+                    return res;
+                }
+            }
+        }
+        sleep(sleep_interval).await;
```

Comment:
Why is this sleep needed? It looks like we can arrive here only when `handles.next()` returns a non-error, which happens when one of the CDC log readers finishes reading. Won't it unnecessarily delay shutdown?

Reply:
Now that I think about it... Is this code even reachable? The upper timestamp is not set for the log readers, so I don't think they can decide to stop by themselves.

Reply:
You're right

Reply:
In fact the loop itself is useless.